firmware/src/app_mqtt.c

```diff
@@ -531,77 +542,96 @@ void handleDownlink(Router__DownlinkMessage* message, void* arg)
-        bw[2]     = sfbw[8];
-        bw[3]     = '\0';
-        bandwidth = atoi(bw);
-    }
-    else
-    {
-        char sf[2];
-        sf[0]           = sfbw[2];
-        sf[1]           = '\0';
-        spreadingfactor = atoi(sf);
-
-        char bw[4];
-        bw[0]     = sfbw[5];
-        bw[1]     = sfbw[6];
-        bw[2]     = sfbw[7];
-        bw[3]     = '\0';
-        bandwidth = atoi(bw);
-    }
-    pkt.datarate = 0x02 << (spreadingfactor - 7);
-
-    if(bandwidth == 500)
-        pkt.bandwidth = LORA_BW_500K;
-    else if(bandwidth == 250)
-        pkt.bandwidth = LORA_BW_250K;
-    else if(bandwidth == 125)
-        pkt.bandwidth = LORA_BW_125K;
-    else
-        pkt.bandwidth = LORA_BW_250K;
+        char* sfbw = lorawan->data_rate;
+        if(sfbw[2] == '1')
+        {
+            char sf[3];
+            sf[0]           = sfbw[2];
+            sf[1]           = sfbw[3];
+            sf[2]           = '\0';
+            spreadingfactor = atoi(sf);
+
+            char bw[4];
+            bw[0]     = sfbw[6];
+            bw[1]     = sfbw[7];
+            bw[2]     = sfbw[8];
+            bw[3]     = '\0';
+            bandwidth = atoi(bw);
+        }
+        else
+        {
+            char sf[2];
+            sf[0]           = sfbw[2];
+            sf[1]           = '\0';
+            spreadingfactor = atoi(sf);
+
+            char bw[4];
+            bw[0]     = sfbw[5];
+            bw[1]     = sfbw[6];
+            bw[2]     = sfbw[7];
+            bw[3]     = '\0';
+            bandwidth = atoi(bw);
+        }
+        pkt.datarate = 0x02 << (spreadingfactor - 7);
+
+        if(bandwidth == 500)
+            pkt.bandwidth = LORA_BW_500K;
+        else if(bandwidth == 250)
+            pkt.bandwidth = LORA_BW_250K;
+        else if(bandwidth == 125)
+            pkt.bandwidth = LORA_BW_125K;
+        else
+            pkt.bandwidth = LORA_BW_250K;
+
+        char*   cr       = lorawan->coding_rate;
+        uint8_t coderate = ((cr[2] - 4) - 0x30);
+        pkt.coderate     = coderate;
+
+        if(gateway_configuration->has_polarization_inversion)
+        {
+            pkt.invert_polarity = gateway_configuration->polarization_inversion;
+        }
 
-    char*   cr       = message->protocol_configuration->lorawan->coding_rate;
-    uint8_t coderate = ((cr[2] - 4) - 0x30);
-    pkt.coderate     = coderate;
+        pkt.preamble  = 8;
+        pkt.no_crc    = true;
+    } else {
+        uint32_t bit_rate            = lorawan->bit_rate;
+        uint32_t frequency_deviation = bit_rate / 2;
 
-    if(message->gateway_configuration->has_polarization_inversion)
-    {
-        pkt.invert_polarity = message->gateway_configuration->polarization_inversion;
-    }
+        if(gateway_configuration->has_frequency_deviation)
+        {
+            frequency_deviation = gateway_configuration->frequency_deviation;
+        }
 
-    pkt.frequency_deviation = message->gateway_configuration->frequency_deviation;
```

Comment:
Shouldn't also be on [this line](https://github.com/TheThingsProducts/gateway/blob/d5498521340ab303cdcda85603081255858f5984/firmware/src/app_mqtt.c#L615)? I'm a bit out of my zone in here but from I could tell the purpose is to maintain the old code if the condition of `lorawan->modulation == LORAWAN__MODULATION__LORA` is true and this is the only line which I didn't see get shifted.

Reply:
No worries - you are correct that it was also set for LoRa modulation before, but not now.

The reason is that frequency deviation is a parameter used only for FSK modulation, not LoRa. The old code probably just tried to make the best use of the API fields, but the frequency deviation only makes sense for FSK.